### ISSUE TITLE
Bug: [@typescript-eslint/no-unused-expressions] False positive with conditional assignments

### ISSUE THREAD

**Issue by Tofandel** [typescript-eslint/typescript-eslint#10323]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.6.3&fileType=.tsx&code=MYewdgzgLgBAFjAvDAjABjQbgLACgA2AprAA5IwDkFOuCAfDGjAGTMwAUZA1Mgl5XAoBKGkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tidmUQAm9AB7FoKZJQ5NUGSImjQO0SODABfEOqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
const h = 100;
let p = '';
h > 0 && (p += h + 'h');
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unused-expressions": ["error"],
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

No error on the third line, which is a conditional assignment (and so not unused)

### Actual Result

Expected an assignment or function call and instead saw an expression. 3:1 - 3:25

### Additional Info

This issue started with v8

**Comment by bradzacher** [comment]
https://eslint.org/docs/latest/rules/no-unused-expressions#allowshortcircuit

**Comment by kirkwaiblinger** [comment]
(was about to say the same)

[Playground](https://typescript-eslint.io/play/#ts=5.6.3&fileType=.tsx&code=MYewdgzgLgBAFjAvDAjABjQbgLACgA2AprAA5IwDkFOuCAfDGjAGTMwAUZA1Mgl5XAoBKGkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tidmUQAm9AB7FoKZJQ5NUGANqQA7gENoTSABpM4bDmXx4HRQGUAFh2j4AwpWhlYlfOjD44iHRAC%2BAXR2eQnkA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

### Code

```ts
const h = 100;
let p = '';
h > 0 && (p += h + 'h');

```

### Eslint config

```json
{
  "rules": {
    "@typescript-eslint/no-unused-expressions": ["warn", {
      "allowShortCircuit": true
    }]
  }
}
```

### TypeScript config

```json
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

| package | version |
| -- | -- |
| `@typescript-eslint/eslint-plugin` | `8.14.0` |
| `@typescript-eslint/parser` | `8.14.0` |
| `TypeScript` | `5.6.3` |
| `ESLint` | `9.14.0` |
| `node` | `web` |
